crates/sui-core/src/epoch/randomness.rs

```diff
@@ -41,6 +42,99 @@ type EncG = bls12381::G2Element;
 
 pub const SINGLETON_KEY: u64 = 0;
 
+// Wrappers for DKG messages (to simplify upgrades).
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[allow(clippy::large_enum_variant)]
+pub enum VersionedProcessedMessage {
+    V0(dkg_v0::ProcessedMessage<PkG, EncG>),
+    V1(dkg_v1::ProcessedMessage<PkG, EncG>),
+}
+
+impl VersionedProcessedMessage {
+    pub fn sender(&self) -> PartyId {
+        match self {
+            VersionedProcessedMessage::V0(msg) => msg.message.sender,
+            VersionedProcessedMessage::V1(msg) => msg.message.sender,
+        }
+    }
+
+    pub fn process(
+        dkg_version: u64,
+        party: Arc<dkg::Party<PkG, EncG>>,
+        message: VersionedDkgMessage,
+    ) -> FastCryptoResult<VersionedProcessedMessage> {
+        match message {
+            VersionedDkgMessage::V0(msg) => {
+                if dkg_version != 0 {
+                    panic!("BUG: message version is 0, expected is {dkg_version}")
+                }
+                let processed = party.process_message(msg, &mut rand::thread_rng())?;
+                Ok(VersionedProcessedMessage::V0(processed))
+            }
+            VersionedDkgMessage::V1(msg) => {
+                if dkg_version != 1 {
+                    panic!("BUG: message version is 1, expected is {dkg_version}")
+                }
+                let processed = party.process_message_v1(msg, &mut rand::thread_rng())?;
+                Ok(VersionedProcessedMessage::V1(processed))
+            }
+        }
+    }
+
+    pub fn merge(
+        dkg_version: u64,
+        party: Arc<dkg::Party<PkG, EncG>>,
+        messages: Vec<Self>,
+    ) -> FastCryptoResult<(VersionedDkgConfimation, VersionedUsedProcessedMessages)> {
+        match dkg_version {
+            0 => {
+                let (conf, msgs) = party.merge(
+                    &messages
+                        .into_iter()
+                        .map(|vm| {
+                            if let VersionedProcessedMessage::V0(msg) = vm {
```

Comment:
rather than having these if let-else panics in the business logic code path, the pattern we've followed in this case is to add methods for `unwrap_v0` and `unwrap_v1`.